modules/imgproc/src/color_hsv.dispatch.cpp

```diff
@@ -274,13 +274,13 @@ bool oclCvtColorBGR2HSV( InputArray _src, OutputArray _dst, int bidx, bool full
 
     if(_src.depth() == CV_8U)
     {
-        static UMat sdiv_data;
-        static UMat hdiv_data180;
-        static UMat hdiv_data256;
-        static int sdiv_table[256];
-        static int hdiv_table180[256];
-        static int hdiv_table256[256];
-        static volatile bool initialized180 = false, initialized256 = false;
+        static thread_local UMat sdiv_data;
+        static thread_local UMat hdiv_data180;
+        static thread_local UMat hdiv_data256;
+        static thread_local int sdiv_table[256];
```

Comment:
Lets try to apply `thread_local` for `UMat` variables only. No need to touch other tables (there is no problem with them).

Use dedicated `initializedUMat` for that.

Reply:
> initializedUMat

I am not sure what you mean by the last part

Reply:
Split initialization on 2 parts:
- regular CPU buffers (like `int sdiv_table[256];`) - they don't need `thread-local` at all.
- OpenCL related stuff which causes problems and needs thread-local variables (also put it under condition that this data is really needed - CPU-only processing doesn't need that at all).

Reply:
probably, even better solution would be to merge all 3 tables together (of course, 3 pointers can be initialized to point to different parts of the joint table) and then we can convert it to UMat on-fly:

```
int rgb2hsv_tab[256*3];
volatile bool rgb2hsv_initialized = false;
if (!rgb2hsv_initialized) { ... }
if (use_opencl) {
UMat rgb2hsv_tab = Mat(1, 256*3, CV_32S, rgb2hsv_tab).getUMat();
}
```

yes, it means extra copy, but compared to the processed data it should be a tiny overhead - much easier to handle.

Reply:
I'm not experienced enough to say the roundtrip-lateny will be more than we'd like, but it think so. But I like the safety and simplicity of your approach, so I guess I'll profile it.

Reply:
@opencv-alalek @vpisarev ...how to know who of you to follow in which case? ^^

Reply:
> probably, even better solution would be to merge all 3 tables together (of course, 3 pointers can be initialized to point to different parts of the joint table) and then we can convert it to UMat on-fly:
> 
> ```
> int rgb2hsv_tab[256*3];
> volatile bool rgb2hsv_initialized = false;
> if (!rgb2hsv_initialized) { ... }
> if (use_opencl) {
> UMat rgb2hsv_utab = Mat(1, 256*3, CV_32S, rgb2hsv_tab).getUMat();
> ... // run opencl kernel with rgb2hsv_utab as one of the parameters
> }
> ```
> 
> yes, it means extra copy, but compared to the processed data it should be a tiny overhead, and much easier to handle than static thread-local UMat's.

also... we are in oclCvtColorBGR2HSV which is "guarded" by CV_OCL_RUN. so i guess i can drop  ```if (use_opencl)```

Reply:
I neither like unecessary round-trips nor messing with the memory model, so i came  up with this:


```c++
    if(_src.depth() == CV_8U)
    {
		static std::mutex mtx;
		{
			std::unique_lock<std::mutex> lock(mtx);
			static UMat sdiv_data;
			static UMat hdiv_data180;
			static UMat hdiv_data256;
			static int combined_table[256];
			static volatile bool initialized180 = false, initialized256 = false;
			static volatile bool initialized = hrange == 180 ? initialized180 : initialized256;

			if (!initialized)
			{
				int * const hdiv_table = hrange == 180 ? combined_table : &combined_table[128], hsv_shift = 12;
				UMat & hdiv_data = hrange == 180 ? hdiv_data180 : hdiv_data256;

				combined_table[0] = 0;

				int v = 255 << hsv_shift;
				if (!initialized180 && !initialized256)
				{
					for(int i = 1; i < 256; i++ )
						combined_table[i] = saturate_cast<int>(v/(1.*i));
					Mat(1, 256, CV_32SC1, combined_table).copyTo(sdiv_data);
				}

				v = hrange << hsv_shift;
				for (int i = 1; i < 256; i++ )
					hdiv_table[i] = saturate_cast<int>(v/(6.*i));

				Mat(1, 256, CV_32SC1, hdiv_table).copyTo(hdiv_data);
				initialized = true;
			}

			h.setArg(ocl::KernelArg::PtrReadOnly(sdiv_data));
			h.setArg(hrange == 256 ? ocl::KernelArg::PtrReadOnly(hdiv_data256) :
							 ocl::KernelArg::PtrReadOnly(hdiv_data180));
		}
	}
```

Reply:
Proposed code locks mutex on each call. That should be properly avoided.

Reply:
@kallaballa, let's try the proposed by me approach without use of static UMat's at all.

Reply:
@opencv-alalek double checked pattern and lock_guard 0f290bc
@vpisarev I'll give it a try and provide numbers

Reply:
Test failed e.g. on MacOS: OCL_TEST_P(CvtColor8u32f, RGB2HSV_FULL) { performTest(3, 3, CVTCODE(RGB2HSV_FULL), IPP_EPS); }

I don't have a MacOS setup, but it stands to reason that it is a concurrency issue. So i checked on the memory model guarantees of ```volatile``` and there are none. So I guessed that is the problem. Let's see the test.

@vpisarev implementation without static still pending. want both version working.